### PR TITLE
s3fs: print unmounting hint when the mount point is stale 

### PR DESCRIPTION
**Pull request by adamqqqplay**
### Details
When situations such as out of memory (OOM) occur, s3fs may receive a SIGKILL signal and exit before cleaning up the mountpoint. At this time, a stale mount point will be left, causing an error to be reported the next time s3fs is mounted: Transport endpoint is not connected.

This patch adds a feature: before s3fs mount, check if the stat information of the mountpoint returns ENOTCONN error, if so, it will execute the umount operation to unmount the stale mountpoint.

### CODE REVIEW

File: src/s3fs.cpp [hunk 1]
```diff
@@ -4668,6 +4703,11 @@ static int my_fuse_opt_proc(void* data, const char* arg, int key, struct fuse_ar
             memset(&stbuf, 0, sizeof stbuf);
             set_mountpoint_attribute(stbuf);
 #else
+            // check stale mountpoint
+            if(stat(arg, &stbuf) == -1 && errno == ENOTCONN){
+                S3FS_PRN_WARN("MOUNTPOINT: directory %s is stale, try to unmount it.", mountpoint.c_str());
+                do_umount(mountpoint, true);
```
Comment (ggtakec) [inline src/s3fs.cpp:4709]: I think it should check the return value of this function, and if it fails, call `S3FS_PRN_EXIT` and `return -1`.

**Comment by gaul** [comment]
Do any other filesystems have this magic behavior?  If not, why should s3fs do something different?  Does this problem really need to be solved?

Instead of remounting the filesystem, could s3fs respond to memory pressure by reducing its cache size or using fewer worker threads?

**Comment by ggtakec** [comment]
@gaul Thanks.
Indeed, if the mount point was already mounted, it is impossible to tell whether it was due to s3fs' end processing missing.
Then, on my second thought umounting it from the s3fs start processing is not a good thing to do.
For example, you can think of the case where it is mounted by another program other than s3fs.
I don't think this process should be done at s3fs startup (even if it's an issue with s3fs' finalization).
After all, this umount processing should be done on the calling side of s3fs if necessary.

File: src/s3fs.cpp (hunk 1 above)
Comment (adamqqqplay) [reply src/s3fs.cpp]: Thanks, I fixed it.

**Comment by adamqqqplay** [comment]
> Thanks for the PR. I think this feature is necessary, so it is good change code.
> 
> There are some errors in Github Actions, so please fix them. Also, I have one comment, so please check it.

Thanks for your review. I fixed the Github Actions errors and your comments, please take a look.

**Comment by ggtakec** [comment]
I confirmed your fix. (I think you've fixed all my requests for corrections, but the test is still failing.)

Then, With this added `do_umount` function call, I'd like to discuss whether or not this should be done.

Right now, s3fs doesn't go as far as executing the `umount` operation and exits.(If the mount information already exists, s3fs will fail to start.)
In addition to the current situation, I think that outputting a detailed message is sufficient.
By displaying the command message of `umount` or `fusermount` in the message at that time, we can show the user the procedure to resolve the error.
With this method, it seems that it is possible to respond not excessively.

**Comment by adamqqqplay** [comment]
> I confirmed your fix. (I think you've fixed all my requests for corrections, but the test is still failing.)
> 
> Then, With this added `do_umount` function call, I'd like to discuss whether or not this should be done.
> 
> Right now, s3fs doesn't go as far as executing the `umount` operation and exits.(If the mount information already exists, s3fs will fail to start.) In addition to the current situation, I think that outputting a detailed message is sufficient. By displaying the command message of `umount` or `fusermount` in the message at that time, we can show the user the procedure to resolve the error. With this method, it seems that it is possible to respond not excessively.

Thanks for your suggestion. I think s3fs output a command hint is very good for users, especially now there are many users who don't know how to fix this issue (such as #2176 #2092).

I will push a new patch soon, thanks.

**Review by ggtakec — APPROVED** [review]
Thank you. There is no problem with this fix.

**Comment by ggtakec** [comment]
@gaul I merged this PR. If you still have any problems (message content, etc.), please let me know.

**Pull request merged**